OneCommunityGlobal/HGNRest
# Diego add test coverage

<img width="616" alt="image" src="https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/7878383f-bdd6-4ddf-ad2f-66391cf3a606">

# Description
Adds a jest code coverage script and updates jest.config.js configurations to enforce meeting a code coverage (this is something we will tweak once we have add the missing unit tests in the backend). In addition, adds a pre-push git hook with husky to prevent pushing the code into the git repository if the code coverage threshold is not met.

## Related PRS:
N/A

## Main changes explained:
- Functionality binds to git hooks with husky (pre-push) to run code coverage script on all files supporting API routes
- Code coverage will only cover files in  the following folders: src/controllers, src/helpers, src/routes, src/utilities and src/websockets

## How to test:
1. check into current branch
2. do npm install
3. checkout into a new branch of your own to test this functionality (since we will create a new commit to test this). eg. git checkout -b yourname/test_code_coverage
4. if running on Mac or Linux run: `npm run prepare` (this will make husky files executable)
5. go to `package.json` and delete the `--passWithNoTests` flag from the `test:ci` script:
![Screen Shot 2024-03-22 at 12.21.23](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/5f9d2198-8bc4-41e3-ba5a-e3c3fec2f9ec)
6. remove the `!` symbol at line 4 in `jest.config.js` file:
![Screen Shot 2024-03-22 at 12.27.10](https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/dffb9ff8-d508-46a7-a7cf-f00889b62485)
7. add your changes to the staging area: `git add .`, commit your changes and push the to your own branch.
8. git should fail in pushing your changes to the repository since jest also failed to found any test files.
9. add a new file. e.g. math.js
10. add some logic for testing. e.g.:
`const sum = (a, b) => a + b;
const substract = (a, b) => a - b;
module.exports = { sum, substract };`
11. add a test file: e.g. `math.spec.js`:
```
describe('math module', () => {
  test('ensure a + b returns the correct value', () => {
    expect(sum(1, 2)).toBe(3);
  });

  test('ensure a - b returns the correct value', () => {
    expect(substract(1, 2)).toBe(-1);
  });
});
```
12. add your changes to the staging area: `git add .`, commit your changes and push the to your own branch.
13. git should fail in pushing your changes to the repository since jest also failed because you didn't meet the code coverage threshold.

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/26840798/51c37932-7878-48bb-952e-adf8aa1ef545


## Note:
Code coverage will be enforced when all the missing unit tests in the backend are added. This will help us develop a mindset to enforce Test Driven Development and become more proficient with our code and make the codebase more reliable.
